### PR TITLE
feat: add support for oauth1

### LINKED ISSUE #18: Support social auth with twitter and other oauth providers that require multiple fields
Problem: twitter (and some other providers) requires frontend to pass two fields, existing serialiser supports only `access_token`

- [ ] Investigated needed changes, potentially:
- [ ] Update SocialToken serialiser to include an extra field next to `access_token`
- [ ] Update do_auth method to take extra / other arguments to allow for multi-arg (token/secret) server-side auth
- [ ] Create tests

Also please see my comment below.

### PR DESCRIPTION
**Pull request by swilltec**
Updated SocialTokenObtainSerializer to support OAuth1 and OAuth2 authentication types

- Updated the SocialTokenObtainSerializer to handle both OAuth1 and OAuth2 authentication types, allowing for the exchange of social authentication tokens for Django REST Framework tokens.
- Improved validation of input data to ensure that required fields are provided and that the selected provider and authentication type are valid choices.
- Refactored the serializer to use constants defined in a separate `constants.py` file for the valid provider choices, promoting code organization and reusability.

## Response from endpoint using a live test twitter tokens
Screenshot 2023-05-30 at 21-10-39 Twitter _ Authorize an application

Screenshot 2023-05-30 at 21-16-07 Social O Auth1 Token List – Django REST framework

closes #18

### CODE REVIEW

File: df_auth/drf/serializers.py [hunk 1]
```diff
@@ -126,27 +128,70 @@ class FirstLastNameSerializerMixin(serializers.Serializer):
...
-    provider = serializers.ChoiceField(
-        choices=[
-            (backend.name, backend.name)
-            for backend in AUTHENTICATION_BACKENDS
-            if hasattr(backend, "name")
-        ],
+    oauth_token = serializers.CharField(write_only=True, required=False)
+    oauth_token_secret = serializers.CharField(write_only=True, required=False)
+    access_token = serializers.CharField(write_only=True, required=False)
+    oauth_type = serializers.ChoiceField(
```
Comment (eugapx) [inline df_auth/drf/serializers.py:134]: I think we can compute `oauth_type` by `provider`. So, we can remove this field.

Also, looks like we don't need all content from `contants.py` then

Comment (swilltec) [inline df_auth/drf/serializers.py]: That can also work. I implemented this way for flexibility, making it easy to switch between different OAuth type for same provider.

File: df_auth/drf/serializers.py [hunk 2]
```diff
@@ -126,27 +128,70 @@ class FirstLastNameSerializerMixin(serializers.Serializer):
...
+        choices=OAuthType.choices, default=OAuthType.OAUTH2, required=False
     )
+    provider = serializers.ChoiceField(choices=PROVIDER_CHOICES)
```
Comment (eugapx) [inline df_auth/drf/serializers.py]: You can limit choices to `OAUTH2_BACKENDS` and `OAUTH1_BACKENDS` here

File: df_auth/drf/serializers.py [hunk 3]
```diff
@@ -126,27 +128,70 @@ class FirstLastNameSerializerMixin(serializers.Serializer):
 
 
 class SocialTokenObtainSerializer(FirstLastNameSerializerMixin, TokenCreateSerializer):
```
Comment (eugapx) [inline df_auth/drf/serializers.py]: OAuth1 and OAuth2 have different input params `access_token` and  `oauth_token + oauth_token_secret `. So, it's better to have 2 different serializers (and views) for them. Then `access_token`, `oauth_token` and `oauth_token_secret` will be required fields. 

So, lets keep `/social` URL for OAuth2 and add `/social/oauth1` for OAuth1

**Review by eugapx — CHANGES_REQUESTED** [review]
Please, install pre-commit hooks and apply them https://pre-commit.com/

File: tests/settings.py [hunk 4]
```diff
@@ -112,3 +113,9 @@
 EMAIL_USE_SSL = os.environ.get("EMAIL_USE_SSL", "")
 EMAIL_HOST_USER = os.environ.get("EMAIL_HOST_USER", "")
 EMAIL_HOST_PASSWORD = os.environ.get("EMAIL_HOST_PASSWORD", "")
+
+SOCIAL_AUTH_TWITTER_KEY = os.environ.get("SOCIAL_AUTH_TWITTER_KEY", "")
+SOCIAL_AUTH_TWITTER_SECRET = os.environ.get("SOCIAL_AUTH_TWITTER_SECRET", "")
+
+SOCIAL_AUTH_TWITTER_KEY = "EZx4rQKH9bEQjJl8destnj1VJ"
```
Comment (eugapx) [inline tests/settings.py]: Please, remove these keys. 

For local development you can use `.env` file instead

File: df_auth/drf/serializers.py [hunk 5]
```diff
@@ -125,18 +122,30 @@ class FirstLastNameSerializerMixin(serializers.Serializer):
...
-            if hasattr(backend, "name")
-        ],
-    )
-
+class SocialTokenObtainBaseSerializer(
+    FirstLastNameSerializerMixin, TokenCreateSerializer
+):
     response = serializers.JSONField(read_only=True)
 
+    def validate(self, attrs):
```
Comment (eugapx) [inline df_auth/drf/serializers.py:132]: You have code duplication here. The only difference between OAuth1 and OAuth2 is 

```
self.user = request.backend.do_auth(attrs["access_token"], user=user)
```

and 

```
self.user = request.backend.do_auth({
            "oauth_token": data["oauth_token"],
            "oauth_token_secret": data["oauth_token_secret"],
        }, user=user)
```

Move token data obtaining in a separate method and use it in validate base method

File: tests/test_app/tests.py [hunk 6]
```diff
@@ -1,12 +1,144 @@
...
+        # Create request object
+        self.request = APIRequestFactory()
+        self.request.user = self.user
+        self.request.session = {}
+        self.request.social_strategy = DRFStrategy(DjangoStorage, self.request)
+        self.access_token = "test"
+
+        self.serializer = SocialTokenObtainSerializer
+
+    def test_fields(self):
```
Comment (eugapx) [inline tests/test_app/tests.py]: This test seems useless for me. It checks not the logic but the declarations of SocialTokenObtainSerializer

File: tests/test_app/tests.py [hunk 7]
```diff
@@ -1,12 +1,144 @@
...
+        )
+        # Create request object
+        self.request = APIRequestFactory()
+        self.request.user = self.user
+        self.request.session = {}
+        self.request.social_strategy = DRFStrategy(DjangoStorage, self.request)
+
+        self.serializer = SocialOAuth1TokenObtainSerializer
+
+    def test_fields(self):
```
Comment (eugapx) [inline tests/test_app/tests.py]: This test seems useless for me. It checks not the logic but the declarations of SocialTokenObtainSerializer

**Review by eugapx — CHANGES_REQUESTED** [review]
I left coments on PR

**Pull request merged**